net/libfab/libfab.c

```diff
@@ -0,0 +1,2829 @@
+		if (ev.nbe_status == 0 && buf->fb_ev_ep != NULL) {
+			ev.nbe_ep = &buf->fb_ev_ep->fep_nep;
+			libfab_ep_get(buf->fb_ev_ep);
+		}
+	}
+	ma->ftm_ntm->ntm_callback_counter++;
+
+	libfab_buf_fini(buf);
+	M0_ASSERT(libfab_tm_invariant(ma));
+	libfab_tm_evpost_lock(ma);
+	libfab_tm_unlock(ma);
+	m0_net_buffer_event_post(&ev);
+	libfab_tm_lock(ma);
+	libfab_tm_evpost_unlock(ma);
+	M0_ASSERT(libfab_tm_invariant(ma));
+	M0_ASSERT(M0_IN(ma->ftm_ntm->ntm_state, (M0_NET_TM_STARTED,
+						 M0_NET_TM_STOPPING)));
+	ma->ftm_ntm->ntm_callback_counter--;
+}
+
+/**
+ * Completes the buffer operation.
+ */
+static void libfab_buf_done(struct m0_fab__buf *buf, int rc)
+{
+	struct m0_fab__tm    *ma = libfab_buf_tm(buf);
+	struct m0_net_buffer *nb = buf->fb_nb;
+	struct m0_fab__buf   *pas_buf;
+	uint64_t             *ptr;
+
+	M0_ENTRY("b=%p rc=%d", buf, rc);
+	M0_PRE(libfab_tm_is_locked(ma));
+	/*
+	 * Multiple libfab_buf_done() calls on the same buffer are possible if
+	 * the buffer is cancelled.
+	 */
+	if (!fab_buf_tlink_is_in(buf)) {
+		/* Try to finalise. */
+		if (m0_thread_self() == &ma->ftm_poller) {
+			if (buf->fb_length == (sizeof(uint64_t) * 2)) {
+				ptr = (uint64_t *)nb->nb_buffer.ov_buf[0];
+				if (*ptr == FAB_DUMMY_DATA) {
+					ptr++;
+					pas_buf = (struct m0_fab__buf *)(*ptr);
+					libfab_buf_complete(pas_buf, 0);
+				}
+			}
+			libfab_buf_complete(buf, rc);
+		}
+		else {
+			/* Otherwise, postpone finalisation to
+			* libfab_tm_buf_done(). */
+			buf->fb_status = rc;
+			fab_buf_tlist_add_tail(&ma->ftm_done, buf);
+		}
+	}
+}
+
+/**
+ * Increments the ref count of the endpoint.
+ */
+static inline void libfab_ep_get(struct m0_fab__ep *ep)
+{
+	m0_ref_get(&ep->fep_nep.nep_ref);
+}
+
+/**
+ * End-point finalisation call-back.
+ *
+ * Used as m0_net_end_point::nep_ref::release(). This call-back is called when
+ * end-point reference count drops to 0.
+ */
+static void libfab_ep_release(struct m0_ref *ref)
+{
+	struct m0_net_end_point *nep;
+	struct m0_fab__ep       *ep;
+	struct m0_fab__tm       *tm;
+
+	nep = container_of(ref, struct m0_net_end_point, nep_ref);
+	ep = libfab_ep_net(nep);
+	tm = nep->nep_tm->ntm_xprt_private;
+
+	m0_nep_tlist_del(nep);
+	libfab_ep_param_free(ep, tm);
+	
+}
+
+/**
+ * Generate unique key for memory registration.
+ */
+static uint64_t libfab_mr_keygen(void)
+{
+	uint64_t key = FAB_MR_KEY + mr_key_idx;
+	mr_key_idx++;
+	return key;
+}
+
+/**
+ * Read single event from event queue.
+ */
+static int libfab_check_for_event(struct fid_eq *eq)
+{
+	struct fi_eq_cm_entry entry;
+	uint32_t              event = 0;
+	ssize_t               rd;
+
+	rd = fi_eq_read(eq, &event, &entry, sizeof(entry), 0);
+	if (rd == -FI_EAVAIL) {
+		struct fi_eq_err_entry err_entry;
+		fi_eq_readerr(eq, &err_entry, 0);
+		M0_LOG(M0_ERROR, "%s %s\n", fi_strerror(err_entry.err),
+		       fi_eq_strerror(eq, err_entry.prov_errno,
+				      err_entry.err_data,NULL, 0));
+		return rd;
+	}
+
+	return event;
```

Comment:
event is uint32_t.
the return value type of this function is int.

Type mismatch.

Reply:
Corrected